src/protocol/transport_service.rs

```diff
@@ -96,6 +96,167 @@ impl ConnectionContext {
+            self.pending_keep_alive_timeouts.push(Box::pin(async move {
+                tokio::time::sleep(timeout).await;
+                (peer, connection_id)
+            }));
+        }
+
+        tracing::trace!(
+            target: LOG_TARGET,
+            ?peer,
+            ?connection_id,
+            ?self.keep_alive_timeout,
+            last_activity = ?self.last_activity,
+            pending_keep_alive_timeouts = ?self.pending_keep_alive_timeouts.len(),
+            "substream activity",
+        );
+
+        // Wake any pending poll.
+        self.waker.take().map(|waker| waker.wake());
+    }
+}
+
+impl Stream for KeepAliveTracker {
+    type Item = (PeerId, ConnectionId);
+
+    fn poll_next(mut self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Option<Self::Item>> {
+        if self.pending_keep_alive_timeouts.is_empty() {
+            // No pending keep-alive timeouts.
+            self.waker = Some(cx.waker().clone());
+            return Poll::Pending;
+        }
+
+        match self.pending_keep_alive_timeouts.poll_next_unpin(cx) {
+            Poll::Ready(Some(key)) => {
+                // Check last-activity time.
+                let Some(when) = self.last_activity.get(&key) else {
+                    tracing::debug!(
+                        target: LOG_TARGET,
+                        peer = ?key.0,
+                        connection_id = ?key.1,
+                        "Last activity no longer tracks the connection (closed event triggered)",
+                    );
+
+                    // We have effectively ignored this `Poll::Ready` event. To prevent the
+                    // future from getting stuck, we need to tell the executor to poll again
+                    // for more events.
+                    cx.waker().wake_by_ref();
+                    return Poll::Pending;
+                };
+
+                // Keep-alive timeout not reached yet.
+                if when.elapsed() < self.keep_alive_timeout {
+                    let timeout = self.keep_alive_timeout - when.elapsed();
+
+                    tracing::debug!(
+                        target: LOG_TARGET,
+                        peer = ?key.0,
+                        connection_id = ?key.1,
+                        ?timeout,
+                        "keep-alive timeout not yet reached",
+                    );
+
+                    // Refill the keep alive timeouts.
+                    self.pending_keep_alive_timeouts.push(Box::pin(async move {
+                        tokio::time::sleep(timeout).await;
+                        key
+                    }));
+
+                    // This is similar to the `last_activity` check above, we need to inform
+                    // the executor that this object may produce more events.
+                    cx.waker().wake_by_ref();
+                    return Poll::Pending;
+                }
+
+                // Keep-alive timeout reached.
+                tracing::trace!(
```

Comment:
nit: This one seems to be less frequent.
```suggestion
                tracing::debug!(
```